luomu-common/src/addr_pair.rs

```diff
@@ -0,0 +1,227 @@
+use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};
+
+use crate::MacAddr;
+
+use super::{Destination, Source};
+
+/// In network protocol implementations addresses usually comes in pairs: Source
+/// and destination IP address, source and destination ports, etc.
+///
+/// This trait is abstraction of this idea.
+pub trait AddrPair<ADDR> {
+    /// Construct new [AddrPair] with given [Source] and [Destination]
+    /// addresses.
+    fn new(src: Source<ADDR>, dst: Destination<ADDR>) -> Self;
+
+    /// Return source address from address pair.
+    fn source(&self) -> Source<ADDR>;
+
+    /// Return destination address from address pair.
+    fn destination(&self) -> Destination<ADDR>;
+
+    /// Return new address pair with source and destination addresses flipped.
+    fn flip(&self) -> Self;
+}
+
+/// In protocols IP addresses usually appear in pairs: Source and destination
+/// IP. This enum provides such pair for both IPv4 and IPv6 addresses.
+///
+/// Main benefit of using this is to have the protocol version associated with
+/// both addresses at the same time providing ergonomics and a bit less memory
+/// consumption.
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq)]
+pub enum IPPair {
+    /// IPv4 address pair
+    V4 {
+        /// Source IPv4 address
+        src: Source<Ipv4Addr>,
+        /// Destination IPv4 address
+        dst: Destination<Ipv4Addr>,
+    },
+    /// IPv6 address pair
+    V6 {
+        /// Source IPv6 address
+        src: Source<Ipv6Addr>,
+        /// Destination IPv4 address
+        dst: Destination<Ipv6Addr>,
+    },
+}
+
+impl AddrPair<IpAddr> for IPPair {
+    /// Construct a new `IPPair`.
+    ///
+    /// # Panics
+    /// Both [IpAddr] need to be same IP version or the call will panic. Use
+    /// [IPPair::new_v4] or [IPPair::new_v6] instead for safe versions.
+    fn new(src: Source<IpAddr>, dst: Destination<IpAddr>) -> IPPair {
+        match (src.unwrap(), dst.unwrap()) {
+            (IpAddr::V4(src), IpAddr::V4(dst)) => {
+                IPPair::new_v4(Source::new(src), Destination::new(dst))
+            }
+            (IpAddr::V6(src), IpAddr::V6(dst)) => {
+                IPPair::new_v6(Source::new(src), Destination::new(dst))
+            }
+            _ => panic!(
+                "IPPair::new() invalid IP address families provided. src: {:?}, dst: {:?}",
+                src, dst
+            ),
+        }
+    }
+
+    /// Returns the source IP address from pair.
+    fn source(&self) -> Source<IpAddr> {
+        match self {
+            IPPair::V4 { src, .. } => Source::new(IpAddr::from(src.unwrap())),
```

Comment:
Why not just `src.clone()`  (or just return `src` it is `Copy`)

Same applies to IPv6 variant and for `destination()`

Reply:
There's a type conversion from `Ipv4Addr` or `Ipv6Addr` to `IpAddr`.

Reply:
Ah, missed that. Sorry.